paddle/framework/scope.h

```diff
@@ -35,73 +35,33 @@ class Scope;
  */
 class Scope {
  public:
-  /**
-   * @brief Initialize s Scope without parent.
-   */
   Scope() {}
+  ~Scope();
 
-  /**
-   * @brief Initialize a Scope with parent.
-   */
-  explicit Scope(const std::shared_ptr<Scope>& parent) : parent_(parent) {}
-
-  /**
-   * @brief Create Variable
-   *
-   * Create Variable in this Scope. Return the exist one if Variable already
-   * been created.
-   */
-  Variable* CreateVariable(const std::string& name) {
-    auto var = GetVariable(name);
-    if (var) {
-      return var;
-    } else {
-      auto ptr = new Variable();
-      name_to_var_[name] = std::unique_ptr<Variable>(ptr);
-      var_to_name_[ptr] = name;
-      return GetVariable(name);
-    }
-  }
-
-  /**
-   * @brief Get Variable.
-   *
-   * Get Variable from this Scope, this function will recursive find Variable
-   * from it's parent scope. Return nullptr if not found.
-   */
-  Variable* GetVariable(const std::string& name) const {
-    auto it = name_to_var_.find(name);
-    if (it != name_to_var_.end()) {
-      return it->second.get();
-    } else if (parent_ != nullptr) {
-      return parent_->GetVariable(name);
-    } else {
-      return nullptr;
-    }
-  }
-
-  /**
-   * @brief If this scope has a Var named name.
-   *
-   * Find if there is a Variable in this scope and it's parent scope
-   */
-  bool HasVariable(const std::string& name) const {
-    return (name_to_var_.find(name) != name_to_var_.end() ||
-            (parent_ && parent_->HasVariable(name)));
-  }
-
-  std::string GetVariableName(Variable* const var) const {
-    try {
-      return var_to_name_.at(var);
-    } catch (...) {
-      return "";
-    }
-  }
+  // Create a sub-scope. Returns a reference other than a pointer so
+  // to prevent from manual deletion.
+  Scope& NewScope();
+
+  // Create a variable with given name if it doesn't exist.
+  Variable* NewVar(const std::string& name);
+
+  // Create a variable with a scope-unique name.
+  Variable* NewVar();
+
+  // Find a variable in the scope or any of its ancestors.  Returns
+  // nullptr if cannot find.
+  Variable* FindVar(const std::string& name) const;
+
+  // Find the scope or an ancestor scope that contains the given variable.
+  Scope* FindScope(const Variable* var);
 
  private:
-  std::unordered_map<Variable*, std::string> var_to_name_;
-  std::unordered_map<std::string, std::unique_ptr<Variable>> name_to_var_;
-  std::shared_ptr<Scope> parent_{nullptr};
+  // Call Scope::NewScope for a sub-scope.
+  explicit Scope(Scope* parent) : parent_(parent) {}
+
+  std::map<std::string, Variable*> vars_;
+  std::list<Scope*> kids_;
```

Comment:
why should we store kid scope and how can we use them? 
In this PR, it seems that we will have a global default scope and all other scopes will be stored and `deleted` by this Scope if itself is deleted.